### ISSUE TITLE
lets talk

### ISSUE THREAD

**Issue by wepl** [temisu/WHDLoadArchiveFS#3]
Hi temisu,
I would like to use this issue for related discussions.

For archivefs_dirCache() it is important, that the order in which entries are provided makes sure that for files in a subdirectory the directory entry above is provided before. This is required to build the structure in WHDLoad.
Is this the case?
I wonder also if archives are possible which do not contain the subdirectory above. Can you tell me?

**Comment by temisu** [comment]
TL;DR;-version

Currently there are no dir-entries for parent entries, neither existing direntries are ordered in any way. However, the code for handling these existed in the first version and I can merge it back

non TL;DR;-version

Archives very rarely contain any dir-entries. Lha-files only have them for empty directories, Zip-files have them if they are deemed required (non-default attributes). Creating the dir-entries is left then to the responsibility decompressor.

The first first I made had the support for Examine/Exnext which meant it was returning "default" dir-entries for the implied directories.

However, when I refactored away the support for Examine/Exnext I also dropped this code for simplification. Now the entries are exactly like they are in the archive (which is pretty random order).

The code I wrote earlier is not exactly 1:1 fit for dirCache. But it is 90% there. I'll merge it back in and sew that it works as intended.

And since we have this back-channel, there is another interesting thing to talk about. This is for fileCache. This is a very strange corner case but got me thinking: Should I call the callback at all when the file length is 0? we use null for stop-condition but AllocVec/AllocMem(0) return nullptr...

**Comment by wepl** [comment]
I think if there are no dir entries in the archive it is best to let them create by whdload. This will also be the most efficient way. Additionally whdload can mark these entries as virtual which may be of use later.
The fact that entries will not be provided in order is not so nice. But for the moment I think we can leave it that way. If a dir entry will be provided after an entry in this dir whdload will ignore this entry as a dummy has been created for file entry before.

Please deliver also files with length of 0. Such files may be used as flags by installed programs and must be processed. Whdload will also return a valid memory pointer for these. The vfs then just doesn’t need to do anything for them. The pointer returned by whdload is part of a larger structure with meta data for this file.

**Comment by temisu** [comment]
Ok, I wont then do a big refactor. Just to let you know that I can change the ordering of entries or create the virtual entries if need be.

What I'm thinking is that if there are "real" direntries in archive, they are probably for reason (include filenote and/or attributes) maybe I can move them first? This way they would not get overwritten by the virtual entries...

**Comment by wepl** [comment]
I have now implemented the creation of missing dirs.
I checked some archives and it seems unix-lha always includes dirs. Maybe because permissions and owner/group. Amiga lha seems not.
If there are dirs in lha/zip they seem always come before the files in them. Or is this the only the tool to display them which reorders them?

**Comment by temisu** [comment]
The test tool displays them in the order which they are in archive - so most likely you are correct - they are always first if they are present. This is something I have not previously spent too much energy on. There really isn't any official specification on Lha and Amiga zip most likely predates the zip RFC, so we have only the implementations to rely on.

On the positive note I have collection of lha/zip files from different archives. I'll see if there are any exceptions to the rule. Also I think we should mostly cover the amiga-implementations anyway :)

**Comment by temisu** [comment]
I think the existing tooling is buggy, to put mildly about directory entries...

However, I have not yet seen a case where directory entry would be after some file using the path. So I think we are good.

Another topic. I know we discussed a bit about progress indicator. Now when playing with compression I think it would be really useful for fileCache(). maybe we should add another callback of type
void progressIndicator(uint32_t current, uint32_t total):

What do you think?

**Comment by wepl** [comment]
The current progress indicator will count entries during dirCache and files during fileCache. This works good for small files. For large files there will be less updates. I still like to leave that way for now. If all works we can think about improvements.

My actual state is that I have implemented loading the VFS and dirCache. I'm now working on fileCache. After testing this I'm going to provide the updated WHDLoad which you can use for testing.
Sorry for the slow progress (work, family,...).

**Comment by temisu** [comment]
No worries about the timing/progress. I'm currently working on the decompressors and that keeps me entertained for some time... (lha is 75% done, zip is barely started)

And progress indicator is something we can thing of something in the future, I just had an idea how the fileCache progress can be made pretty linear & granular. dirCache progress is almost impossible for archives. For zip that is almost a no-op and for lha there is no idea how entries there are until you have parsed all of them.

**Comment by wepl** [comment]
Hi Teemu,
I have released WHDLoad beta including partial VFS support: http://whdload.de/whdload/whd187dev.lha
You may check the history for details: http://whdload.de/docs/History.html
Regarding release procedure: I think for the moment it is best to only link your git repository from the WHDLoad package and not to include your VFS. What do you think?
Regarding current status: IMHO we could invite people on eab for testing but I don't know what your status is. Is it ready for this?

**Comment by temisu** [comment]
Hi,

Let me quickly check the beta version before we notify others at EAB. It would be sad if there are show stopper bugs.

About the distribution: I have not strong opinion (and even if I did I intentionally put the license as BSD, thus my approval is not needed for distribution. It is as free as it gets - just include the license file). However, I think the link is fine for now. I'm sure it will evolve a lot in the coming months. If it is worth bundling - we see later.

The current status is as follows:
- uncompressed support for lha/zip is complete.
- compressed Lha missing LH1 decompressor (which is very rare and maybe unnecessary)
- Zip support should be complete. there are certain combinations that are not yet covered under the tests.
- The decompression performance is crappy (I frankly blame vbcc code generation) For that I want to try gcc 10 which I managed to create cross compiler for m68k but I'm uncertain if it works. This is something I still need to work on (but does not stop making a pre release)
- Test coverage is not very high. But passable

In short I think it is ready for wider audience. But I need to make a release from the latest state and few tests. So a moment if you will :)

**Comment by wepl** [comment]
Started some tests...
I have updated my repo to your latest commit and now lha-5 and lha-6 are no longer working. Before it worked (zip not before and after). Uncompressed works for lha/zip.
I think I had commit 'Initial round of deflate (full files only so far)' before.
I think LH1 is wasted time, I would not recommend to spend time on this. But an better error message would be useful. It now returns #7 'decompression failure', better would be 'unsupported compression method'.
vbcc code generation is not very well. Maybe it is worth to try other compilers.

**Comment by temisu** [comment]
I tried to prepare everything for public release.

I could not actually reproduce your problem of non-working compressed archives. Maybe it is something on your build? Can you try the binary release I created?

Other than that looking really good! (assuming you also get the decompression working) and we can start sharing this work with others

**Comment by wepl** [comment]
Your build has a different size but also fails with #-7, Decompression failure for lha-[56] and zip-[1-9].
Which VBCC/VASM version do you use? (I have vbcc V0.9g, vasm 1.8h)
Maybe it's the test data. I'm testing BeneathASteelSkyCD32 ca. 100 MiB uncompressed.

**Comment by wepl** [comment]
Small archives ~4MiB are working.

**Comment by wepl** [comment]
08d56bcafab5c21ff15c9948911c083cfea4031e 
did broke it

**Comment by temisu** [comment]
Thanks for the triage, it is very specific :)

I did use smaller files to test and I did not see this issue yet.

I'll repro and fix the issue...

**Comment by wepl** [comment]
Do you want the archive for testing?
You can download the archive here: http://whdload.de/tmp/data-cd32.5.lha

**Comment by temisu** [comment]
I already managed to create a test-archive.

Now it should be easy. I managed to repro it in PC and I'm single stepping it in debugger

Thanks

**Comment by temisu** [comment]
There were actually multiple issues - some harder to pinpoint than other. Those should be now fixed.
BASS install seems to be a good test case as well ;)

Please see if you can make it work now. I had success now both with lha and zip.

While waiting several minutes for it to load it got me thinking that there is a perfect feature we might want to add later.
I'm thinking about preloading the files as compressed stream. This would.
* Reduce the memory required to preload stuff
* Improve the loading time of the install.

Obviously this would only make sense for those cases where memory is limited. And the APIs would need a bit improvement. But other than that what do you think?

**Comment by wepl** [comment]
Thanks, it's working fine now.
I made some tests:

```
CSPPC/68060@60, IBM-DNES-318350W
OS 3.1.4.1 SFS 1.279
XPK: xpackbest mash.100 sqsh.100

BeneathASteelSkyCD32
-mode-  ---size---  -time-
plain	90.441.369   36.12
xpk	58.236.051   60.86
lha-0	90.727.466   44.86
lha-5	66.901.471  203.64
lha-6   66.962.830  204.32
zip-0	91.136.817   43.76
zip-1	68.301.169  254.54
zip-9	67.191.970  246.92

SpaceAce2
plain	 5.121.056    0.40  0.38
xpk	 3.840.256    2.04  2.04
lha-0	 5.120.341    2.16  2.18
lha-5	 3.679.807   11.28 11.26
lha-6    3.638.673   10.92 10.90
zip-0	 5.120.926    2.16  2.18
zip-1	 3.757.776   13.48 13.48
zip-9	 3.644,878   13.12 13.12

```
I had the hope that uncompressed lha/zip would be maybe faster than plain files.
Lha seems to be faster than zip.

Can you explain 'preloading the files as compressed stream'?

**Comment by temisu** [comment]
Hi,

Your performance testing quantifies exactly why I'm worried about performance But when I see stuff like

`
x=~/value
`

compiled to

`
     d86:       7000            moveq #0,%d0
     d88:       3004            movew %d4,%d0
     d8a:       4680            notl %d0
     d8c:       3800            movew %d0,%d4
`

It is clear that there is no amount of source level optimization that would get this somewhere comparable to assembly optimised decompressors. It is already losing it by factor of 4. And I'm using the latest public vbcc / vasm combination - It just does not have any good arithmetic code generation capability - this is what it does (looking the source)

Now when we have a working release, test files and thanks to you the performance test I will continue if I can get gcc do better (Well, the code generation is better but it does have its own set of problems). I think it is good for first unoptimised release if it works...

I think the non-compressed bad performance is coming from a slow memcpy (oneliner that I wrote quickly). This should be fixable but it has not been on the top of my todo yet. I'll see if the vbcc libs have something better and if not - I can experiment with movem-based variant, that should do the trick. (I already have that code, but it is written using gcc inline assembly which does not really mix with vbcc, so I'd rather do the compiler switch first)

Anyway, if it is really working that would be great. Do you think it is worth sharing with other with the note that performance sucks?

Also, about my idea (I have plenty of those). I was thinking that maybe for memory-constrained systems it would be useful to load the compressed files as-is into memory (when there is not enough memory to cache uncompressed files) And then decompress them when they are needed (The code already supports partial reads from any offset). This would be tradeoff where initial load would be faster but reading files would be slower (Probably they are not read all at once so maybe it would not be too disturbing). But if there is not enough memory to cache uncompressed files we would anyway save the switch to OS. This idea is still rather raw (It would require some new APIs) but it is technically doable...

**Comment by temisu** [comment]
Well, the bad performance on uncompressed reads has been addressed (I hope). Try now if you can get better numbers

**Comment by wepl** [comment]
If you have specific complains regarding the generated code of VBCC you may write Frank Wille. He was mostly very open to my improvement requests on VASM. Maybe he can also do something on VBCC. Or maybe can give source level recommendations?

That lha/zip cannot beat xpk was clear. These archivers were never designed for fast decompression. This was also my initial reserve to use them. But for interoperability they are good. And I also see the use case to use them uncompressed to avoid having a lot of files on the disk.
For me the performance for compressed files is more or less as expected.

Your last patch gives a boost for uncompressed lha/zip. With SpaceAce2 it's now equal to plain and with BASS it faster than plain. But my performance test currently does not count the initialize call which takes significant time on large archives.
I also tried to create a PR for my test script (adding 2 files). Don't know if you saw it. But it contains all there merges and other stuff. I wanted to create a PR only for the last commit, but didn't succeed.

I think it is ready and worth to give it to testers. Would recommend to create a new release.
About performance, I think this will depend on machine specs. It doesn't sucks! Many games are small. Even if it loads slower this will be not a problem. BASS-CD32 is surely not a normal use case.
Feel free to post on EAB about it :-)

About your idea to load the data compressed and decompress on demand, this was also what I had in mind in the beginning. For the moment it looks too complex too me. It would surely have big benefits for large games where a lot of data gets probably never loaded. The downside would be that loading at runtime would be slower (as you already wrote). I would like to finish all remaining open points (e.g. loading on demand) before enhancing in this direction.

There is also an improvement possible regarding progress display. Currently WHDLoad does not know when a file is decompressed due current API. It will increase the progress when it gives the memory via allocFile back to wvfs_fileCache. So before the file is decompressed. Not a big thing. We can address this later.

**Comment by temisu** [comment]
Few things.

First. I gave you write access to the repo - no need to for you to do PRs anymore. Just push.
I also saw your PR and approved it but could not merge it since it was done as a draft. Now you can merge it by yourself

Definitely there are further work which we can do. progress bar and compressed loading would be the main candidates.
For progress bar display there already is a some preliminary stuff on the library. You can try the command line application with progress <filename> to see it yourself.

In any case I'll let EAB know. And we can see if there is some traction to this feature. For now I'm closing this thread (lets open a new one when needed) Probably there is going to be some bugs next
